### PR TITLE
[POC] Feature/13035 aws s3 support

### PR DESCRIPTION
**Pull request by Tsukilc**
### 1. Changes

This PR mainly introduces support for S3 in BanyanDB, enabling remote storage snapshot backups. The specific changes include the following:

- **New S3 Remote Storage Support**: Implemented `s3FS`, providing `upload`, `download`, `delete`, and other interfaces. It interacts with S3 using the AWS SDK for Go.
- **Introduced** `aws/config.go`: A configuration class for storing AWS-related information such as bucket name, credentials, and region, without modifying the existing interfaces.
- **Testing and Local Validation**: 

- Verified the API calls locally using S3.
  - Future plans to add unit tests to handle edge cases.

- **Basic Error Handling**: Includes parameter validation.

### 2. Run Screenshots
image
image

### 3. Future Plans

This PR only implements basic S3 functionality. Future plans include:

- **Bucket Management Optimization**: 

- Currently, separate buckets are created for each data type (Measure, Stream, Property). Each data type should have a different management approach. Further study will help us provide more customizable interfaces to users.
  - Allow users to customize bucket names, providing flexible configuration options.

- **Error Handling**: 

Plan to simulate various user behaviors, such as network interruption, repeated startups, high latency, and lack of permissions, and improve error handling to enhance user experience.

- **Expansion Compatibility**: 

Currently supports AWS S3. In the future, we plan to improve extensibility to support other object storage systems like MinIO, Aliyun OSS, etc.

- **Usage Documentation**: 

Add easy-to-understand usage documentation.

### 4. Possible Issues

1. **Bucket Storage**: To implement separate storage for three types of data, AWS upload needs to know the bucket name. However, the current `remote.FS` interface cannot pass this parameter.

- **Solution**: Introduced `catalog.go` as a state class. Before entering the FS interface method, the current data type (Measure, Stream, etc.) is saved in the `NowCatalog` object within `catalog.go`.

2.  **Bucket Storage**: The three data types differ greatly, which may require independent permissions or lifecycle management, but this increases storage and management costs.
3.  **Extensibility**: Consider maintaining the extensibility of the AWS protocol in the future, allowing for the integration of MinIO, OSS, and other object storage systems.

### CODE REVIEW

**Comment by wu-sheng** [comment]
fyi @hanahmily

**Comment by wu-sheng** [comment]
@hanahmily Should we add storing implementation as well? I think it should make more sense. Otherwise, who is going to download these snapshots from data nodes?

**Comment by hanahmily** [comment]
> @hanahmily Should we add restoring implementation as well? I think it should make more sense. Otherwise, who is going to download these snapshots from data nodes?

Yes, we should. I will review it. If all goes well, https://github.com/apache/skywalking-banyandb/blob/main/banyand/backup/restore.go will be improved as well.

**Comment by wu-sheng** [comment]
restore and backup should be paired. Yoi could continue to work on this.

**Comment by Tsukilc** [comment]
> restore and backup should be paired. Yoi could continue to work on this.

Got it! I'll continue working on this. Thanks for the guidance！

File: banyand/backup/backup.go [hunk 1]
```diff
@@ -159,6 +195,9 @@ func newFS(dest string) (remote.FS, error) {
 	switch u.Scheme {
 	case "file":
 		return local.NewFS(u.Path)
+	case "s3":
+		return aws.NewFS()
```
Comment (hanahmily) [inline banyand/backup/backup.go:199]: It should accept u.Path as its base path.

**Review by hanahmily — CHANGES_REQUESTED** [review]

**Comment by hanahmily** [comment]
@Tsukilc Could you use https://github.com/minio/minio to test your AWS S3 FS since it's a Go-based and S3-compatible library?.

**Comment by Tsukilc** [comment]
> @Tsukilc Could you use https://github.com/minio/minio to test your AWS S3 FS since it's a Go-based and S3-compatible library?.

Thanks for your suggestion! I will add tests using minio

File: pkg/fs/remote/catalog.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package remote
+
+// NowCatalog is the catalog name for the current remote.
+var NowCatalog string
```
[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/catalog.go:21]: If we need to divide the measure, stream, and property into three separate buckets, passing the bucket name to newFS would create three different FS instances. The solutions I can think of are:
1. Modify the remote.FS interface to include a bucketName parameter.
2. Or, parse the path to extract the type (e.g., /2025-03-16/measure), and derive the catalog (measure) through string manipulation. The downside is that this approach is tightly coupled with the path format.
3. Or, do we actually not need to separate into three buckets?
My idea is to try not to change the previous code. If it is convenient for you, could you help me to explain further how to deal with this problem? thank you！

[the remark this answers is not in the document]
Comment (hanahmily) [reply pkg/fs/remote/catalog.go:21]: As seen in your uploaded screenshot, we have a "catalog" within the "timerdir"(2025-03-16/stream).  Feel free to remove the catalog prefix.

[the remark this answers is not in the document]
Comment (hanahmily) [reply pkg/fs/remote/catalog.go:21]: BTW, Please sync your local code with the main branch. I have removed the "data" segment from the remote path.

[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/catalog.go:21]: Thank you! I will parse the "catalog" from the "timerdir"

[the remark this answers is not in the document]
Comment (hanahmily) [reply pkg/fs/remote/catalog.go:21]: Let me clarify the bucket strategy: 

If the "dest" flag is set to "s3://banyandb-remote," then "banyandb-remote" is the bucket. We do not need to group files into catalog-based buckets; a single bucket is good for our needs.

[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/catalog.go:21]: Thank you! Now I understand how to do it

File: banyand/backup/backup.go [hunk 3]
```diff
@@ -196,8 +160,7 @@ func newFS(dest string) (remote.FS, error) {
 	case "file":
 		return local.NewFS(u.Path)
 	case "s3":
-		return aws.NewFS()
-
+		return aws.NewFS(dest)
```
Comment (hanahmily) [inline banyand/backup/backup.go:163]: You should use `u.Path` which is the result of `url.Parse(dest)`

File: banyand/backup/backup.go (hunk 1 above)
Comment (Tsukilc) [reply banyand/backup/backup.go]: [062b68a](https://github.com/apache/skywalking-banyandb/pull/622/commits/062b68a57bcd9d39436b6652156a144c6897949e)

File: pkg/fs/remote/aws/aws.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	return &s3FS{
+		client: client,
+	}, nil
+}
+
+func (s *s3FS) Upload(ctx context.Context, path string, data io.Reader) error {
+	_, err := s.client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: getBucketName(),
+		Key:    &path,
+		Body:   data,
```
[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/aws/aws.go]: Improved in commit [ca76162](https://github.com/apache/skywalking-banyandb/pull/622/commits/ca761620a613807b02a0637f2168f842abca94a5)

File: pkg/fs/remote/aws/config.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package aws
+
+import "time"
+
+// S3Config holds the configuration for S3 storage.
+type S3Config struct {
```
[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/aws/config.go]: I resolve it in [57c2abb](https://github.com/apache/skywalking-banyandb/pull/622/commits/57c2abbf905b89a8825e0cef120d3f4e2cea483a)，The default profile is called banyandb
image

File: pkg/fs/remote/aws/aws.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	awsCfg, err := config.LoadDefaultConfig(
+		context.TODO(),
+		config.WithRegion(cfg.Region),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
+			cfg.KeyID,
+			cfg.SecretKey,
+			"",
+		)),
+		config.WithHTTPClient(httpClient),
+		config.WithClientLogMode(aws.LogRetries),
```
[the remark this answers is not in the document]
Comment (Tsukilc) [reply pkg/fs/remote/aws/aws.go]: I checked the official documentation and found that the AWS SDK for Go has a default exponential backoff strategy. Users can configure the maximum number of retries through the shared configuration file. 
image
image

**Comment by Tsukilc** [comment]
I removed one test case in [ebcb63a](https://github.com/apache/skywalking-banyandb/pull/622/commits/ebcb63a40bf444aa7ca05f4b71a9ae32ca08ce9e) because newFS needed to read the shared profile, but the ci environment couldn't provide s3's public and private keys like local tests

**Comment by wu-sheng** [comment]
How about using MinIO to mock that remote service?

**Comment by Tsukilc** [comment]
> How about using MinIO to mock that remote service?

Thanks for the suggestion!  I will try to use it

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -69,7 +69,56 @@ jobs:
         uses: actions/setup-go@v3
         with:
           go-version-file: 'go.mod'
-      - name: Update dependencies 
+
+      - name: Set up Docker Buildx
```
Comment (wu-sheng) [inline .github/workflows/test.yml:73]: To @hanahmily This test seems very related to s3 only. We should have separate GHA file to control and verify this case.

Comment (Tsukilc) [reply .github/workflows/test.yml]: Yes! I tried to extract it in [b3992b8](https://github.com/apache/skywalking-banyandb/pull/622/commits/b3992b8f09e71e0bdabecadbb0d8834149e2b80f),moved it to.github/actions/s3-setup/action.yml

> To @hanahmily This test seems very related to s3 only. We should have separate GHA file to control and verify this case.

**Comment by wu-sheng** [comment]
Could you try to run this GHA into your fork? I think the PR wouldn't run it until this file gets merged in the first time.

**Comment by Tsukilc** [comment]
> Could you try to run this GHA into your fork? I think the PR wouldn't run it until this file gets merged in the first time.

ok, I'm already trying this, There may also be adjustments to the file later

**Comment by wu-sheng** [comment]
You can see no test result of your case.

**Comment by Tsukilc** [comment]
yes, but I see a new one I added to the test workflow called call s3 setup, maybe it works?

**Comment by wu-sheng** [comment]
I don't see it on https://github.com/apache/skywalking-banyandb/actions or PR CI list.

**Comment by Tsukilc** [comment]
Sorry, I misunderstood. It should be a new workflow in PR CI list, not mixed in with test.yml. 
My current idea is that the S3 tests should only run on GitHub Actions, and not be included when running 'make test' locally(Otherwise, when local developers make test, they also need to configure minio to pass the test).
 I'll test it in my forked repository.

**Comment by wu-sheng** [comment]
I think you are still not following my words. I mean in this PR, we can't see this new PR running.
image

I mean you can't see it in here.

**Comment by Tsukilc** [comment]
Thanks! the final behavior might be like this picture? 
  I misunderstood earlier that the workflow didn’t include test-s3 in [b3992b8](https://github.com/apache/skywalking-banyandb/pull/622/commits/b3992b8f09e71e0bdabecadbb0d8834149e2b80f) because of coding problems. 
The actual behavior is that adding the test-s3 file in github.workflows will trigger the workflow in this PR (experiment done in my forked repository).
image

**Comment by wu-sheng** [comment]
Yes, that happens only when this is merged. But the point of PR review is to validate things before merging, right?

**Comment by wu-sheng** [comment]
If this file is OK for @hanahmily , I would recommend you to add this file(exclude the test, but include the installation pre-steps) through another PR, first, then, we back to this PR continue the following works.

**Comment by Tsukilc** [comment]
yes! thanks for your instruction!  But I saw this effect in my fork repository without merging the pr. In this case, the workflow not updated because my file was written incorrectly

**Comment by Tsukilc** [comment]
I'm writing a test on s3 in my fork repository, trying to make sure the  ci file's fully usable after I submit here

**Comment by wu-sheng** [comment]
> yes! thanks for your instruction! But I saw this effect in my fork repository without merging the pr. In this case, the workflow not updated because my file was written incorrectly

I feel otherwise even your file is correct. From my understanding, GH blocks that run in PR CI pipeline(in upstream INFRA in this case) on purpose.

**Comment by Tsukilc** [comment]
I guess I misunderstood, and it turned out to be a happy accident. The purpose of my commit was originally just to refactor the S3-related code, not to be part of the workflow

**Comment by hanahmily** [comment]
@Tsukilc 

Please move your test case to https://github.com/apache/skywalking-banyandb/tree/main/test/integration/standalone/. You don't need to add a new GitHub Actions workflow.

Refer to https://github.com/apache/skywalking-banyandb/tree/main/test/integration/standalone/backup for guidance in creating a "backup-s3". You can copy most of the test setup logic from the backup. The only additional step required is to set up the MinIO server (which you may need at https://github.com/ory/dockertest) before proceeding to https://github.com/apache/skywalking-banyandb/blob/main/test/integration/standalone/backup/backup_suite_test.go#L74.

Test cases need adjustments because they rely on local files.

I believe there are some details regarding the testing that we need to discuss. I highly recommend that you join the Skywalking Slack channel so we can discuss them directly.

**Comment by hanahmily** [comment]
The overall implementation looks good to me. However, the comments on this PR are too long to review. Could you please close it and start a new one to add test cases?
